### ISSUE TITLE
get_class not working for MultiContainerInterfaces

### ISSUE THREAD

**Issue by angevineMiller** [NeurodataWithoutBorders/pynwb#888]
We would like to create MultiContainerInterface extensions for the Frank Lab. We have gotten them working using the pattern described in the [Extensions tutorial](https://pynwb.readthedocs.io/en/latest/tutorials/general/extensions.html#creating-and-using-a-custom-multicontainerinterface), which utilizes nested NWBGroupSpecs. However, we would like to just call [get_class](https://pynwb.readthedocs.io/en/latest/pynwb.html#pynwb.get_class) to get a working MultiContainerInterface class instead of having to write the Python classes ourselves.

We were able to get nested groups working, thanks to @bendichter et al. in (https://github.com/NeurodataWithoutBorders/pynwb/issues/522), but we have not gotten MultiContainerInterfaces to work with get_class yet.

```python
# ------------------
# Create the spec
# ------------------
import pynwb
from pynwb import load_namespaces
from pynwb.spec import NWBNamespaceBuilder, NWBGroupSpec

name = 'multicontainer'
ns_path = name + ".namespace.yaml"
ext_source = name + ".extensions.yaml"

potato = NWBGroupSpec(neurodata_type_def='Potato',
                      neurodata_type_inc='NWBDataInterface',
                      doc='time of multicontainer', 
                      quantity='*'
                      )

potato_sack = NWBGroupSpec(neurodata_type_def='PotatoSack',
                           neurodata_type_inc='NWBDataInterface',
                           name='potato_sack',
                           doc='test of multicontainer', 
                           groups=[potato]
                           )

ns_builder = NWBNamespaceBuilder(name + ' extensions', name)
ns_builder.add_spec(ext_source, potato_sack)
ns_builder.export(ns_path)

# ------------------
# Load and try to use the spec
# ------------------
load_namespaces(ns_path)
Potato = get_class('Potato', 'multicontainer')
PotatoSack = get_class('PotatoSack', 'multicontainer')

my_potatos = PotatoSack(name='my_potatos', 
                        potatos=[Potato(name='One Potato'), 
                                 Potato(name='Two Potato')])

>> TypeError: incorrect type for 'potatos' (got 'list', expected 'Container')
```

Perhaps as expected, we see that PotatoSack is a NWBDataInterface, not a MultiContainerInterface. We then tried changing the neurodata_type_inc of PotatoSack to MultiContainerInterface, this also resulted in an NWBDataInterface, not a MultiContainerInterface.

- **Python Executable:** Conda
- **Python Version:** Python 3.6.7
- **Operating System:** macOS
- **Pynwb Version:** 1.0.1.post0.dev19

## Checklist

- [x] Have you ensured the feature or change was not already [reported](https://github.com/NeurodataWithoutBorders/pynwb/issues/) ?
- [x] Have you included a brief and descriptive title?
- [x] Have you included a clear description of the problem you are trying to solve?
- [x] Have you included a minimal code snippet that reproduces the issue you are encountering?
- [x] Have you checked our [Contributing](https://github.com/NeurodataWithoutBorders/pynwb/blob/dev/docs/CONTRIBUTING.rst) document?

**Comment by bendichter** [comment]
@angevineMiller 

Thanks for raising this issue. In fact, this was next on my list! It would be trivial if `MultiContainerInterface` were a neurodata_type and you could automatically inherit from it by listing it as your `neurodata__type_inc`, however it is not a neurodata type, and it has been decided [here](https://github.com/NeurodataWithoutBorders/nwb-schema/issues/257) that it should not become one. 

In order to get this working, we'll need to create some additional code to check if the neurodata_types contained within are of quantity '*', and then automatically make the generated class inherit from `MultiContainerInterface` and add the appropriate attributes to automatically generate the appropriate methods.

**Comment by bendichter** [comment]
Since the meat of `get_class` is in HDMF, I think this will require `MultiContainerInterface` to move over there or we'll get a circular import error (`self.__get_container_cls(namespace, 'MultiContainerInterface')` would only work if `MultiContainerInterface` were a ndtype). `MultiContainerInterface` currently inherits from `NWBDataInterface`, so a repo transfer will require a bit of reworking class inheritance.

**Comment by angevineMiller** [comment]
Thanks for the response @bendichter.  

I agree that if ```MultiContainerInterface``` is not going to become a neurodata_type, but rather is just viewed as different API functionality for the same underlying ```NWBDataInterface```, then we probably want ```get_class``` to infer whether it needs to return a ```MultiContainerInterface```.  As you suggested, perhaps we could do this by checking whether any of the contained neurodata_types have quantity other than '1' or '?'. We also want to handle the case where only some, but not all, of the contained neurodata_types are need ```MultiContainerInterface``` functionality.

Separately, do we want to handle neurodata_types with a discrete quantity greater than 1, such as '4' (e.g. the 4 wheels of a car)?  Would we throw an error if more than 4 of these are added to the MultiContainerInterface?

**Comment by bendichter** [comment]
Good question! I had not thought of that. I don't think we have any such cases but it wouldn't be hard to track that functionality on the MultiContainerInterface

**Comment by oruebel** [comment]
> I think this will require `MultiContainerInterface` to move over there

Moving the MultiContainerInterface class to HDMF I think makes sense. My guess is that there are probably also a couple of other general classes (ie., support classes that are not specific to NWB)  in core.py that would make sense to move (e.g, LabelledDict)

> We also want to handle the case where only some, but not all, of the contained neurodata_types are need `MultiContainerInterface` functionality.

Having a simple logic in get_class that determines if MultiContainerInterface should be used makes sense. However, for this sort of customization of behavior (where you want it apply for some, but not all cases), I think we should require that the user write their own class. In particular, another user will not know if and how they should customize get_class behavior to use a particular extensions. Also on read, the I/O backend needs to be able to use get_class to automatically generate a container.   

> Separately, do we want to handle neurodata_types with a discrete quantity greater than 1

Since this should be simple to do and we have a well-defined reason, it's fine to add this functionality. However, since we currently don't have a specific case that requires this, I would view this as not critical.

**Comment by tjd2002** [comment]
xref #556 ?

**Comment by bendichter** [comment]
fixed by hdmf#420
